### PR TITLE
Add exception handling for bloom option json reading

### PR DESCRIPTION
**Pull request by elliVM**
- Catch exceptions thrown by gson when reading filter fields option string to JSON
- Use FilterField object to store filter field options `expected` and `fpp` and test
- Get BloomFilter `bitSize` from FilterField object
- Add tests for incorrect JSON fromat in options and duplicate filter field values

### CODE REVIEW

File: src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java [hunk 1]
```diff
@@ -162,32 +158,35 @@ private Dataset<Row> estimateSize(Dataset<Row> dataset) {
     }
 
     public Dataset<Row> aggregate(Dataset<Row> dataset) {
-
-        FilterTypes filterTypes = new FilterTypes(this.zeppelinConfig);
-
-        BloomFilterAggregator agg = new BloomFilterAggregator(inputCol, estimateCol, filterTypes.sortedMap());
-
+        final FilterTypes filterTypes = new FilterTypes(this.zeppelinConfig);
```
Comment (51-code) [inline src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java:161]: I would have this FilterTypes object as a class variable rather than the zeppelinConfig. Realistically though, if thinking with single responsibility principle in mind, this object should be four objects. Now it's resposnsible for creating, updating, estimating and aggregating (like in the Enum). If it was refactored into multiple objects, the FilterTypes as class variable suggestion would make more sense too. If that sounds good, I would open a new issue and eventually another PR for that so this one doesn't get too big.

File: src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java [hunk 2]
```diff
@@ -162,32 +158,35 @@ private Dataset<Row> estimateSize(Dataset<Row> dataset) {
...
+        final SortedMap<Long, Double> map = new TreeMap<>();
+        final List<FilterField> fieldList = filterTypes.fieldList();
+        for (final FilterField field : fieldList) {
+            map.put(field.expected(), field.fpp());
+        }
+        final BloomFilterAggregator agg = new BloomFilterAggregator(inputCol, estimateCol, map);
```
Comment (51-code) [inline src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java:167]: I like the idea of having the FilterTypes as a list but it feels weird that then it has to be put into a map instantly anyways like before. Could the BloomFilterAggregator be refactored to work with the list too?

File: src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterField.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * Teragrep, the applicable Commercial License may apply to this file if you as
+ * a licensee so wish it.
+ */
+package com.teragrep.pth10.steps.teragrep.bloomfilter;
+
+import org.apache.spark.util.sketch.BloomFilter;
+
+import java.util.Objects;
+
+public class FilterField {
```
Comment (51-code) [inline src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterField.java:52]: Final keyword missing in class declaration.

File: src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterTypes.java [hunk 4]
```diff
@@ -73,35 +72,33 @@ public FilterTypes(Config config) {
...
-                throw new RuntimeException("JSON did not have expected values of 'expected' or 'fpp'");
-            }
+    public List<FilterField> fieldList() {
+        final List<FilterField> fieldList;
+        try {
+            final Gson gson = new Gson();
+            fieldList = gson.fromJson(sizesJsonString(), new TypeToken<List<FilterField>>() {
+            }.getType());
+        }
+        catch (Exception e) {
```
Comment (51-code) [inline src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterTypes.java:82]: Would it be possible to catch a more specific exception, like JsonSyntaxException (I'm not 100% sure what this fromJson() throws)

File: src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterTypes.java [hunk 5]
```diff
@@ -73,35 +72,33 @@ public FilterTypes(Config config) {
...
+            throw new RuntimeException("Found duplicate values in 'dpl.pth_06.bloom.db.fields'");
         }
-        return sizesMapFromJson;
+        return fieldList;
     }
 
     public Map<Long, Long> bitSizeMap() {
-        final Map<Long, Double> filterSizes = sortedMap();
+        final List<FilterField> fieldList = fieldList();
+        System.out.println(fieldList);
```
Comment (51-code) [inline src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterTypes.java:97]: A debug print is left in the code here.

**Review by 51-code — CHANGES_REQUESTED** [review]
Requested some changes.

File: src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java (hunk 1 above)
Comment (elliVM) [reply src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java:161]: Good suggestion I am looking into it but I also created and issue about splitting this object https://github.com/teragrep/pth_10/issues/374

File: src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java (hunk 2 above)
Comment (elliVM) [reply src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java:167]: Indeed a bit unfortunate result of refactoring but refactored BloomFilterAggregator would require new release of dpf_03. Also would couple dpf_03 to this custom object type a general map might be better for use elsewhere.

File: src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterTypes.java (hunk 4 above)
Comment (elliVM) [reply src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterTypes.java]: catching exceptions `JsonIOException` and `JsonSyntaxException` thrown by `fromJson()` now

File: src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterTypes.java (hunk 5 above)
Comment (elliVM) [reply src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterTypes.java]: removed!

File: src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterField.java (hunk 3 above)
Comment (elliVM) [reply src/main/java/com/teragrep/pth10/steps/teragrep/bloomfilter/FilterField.java]: made final

File: src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java (hunk 1 above)
Comment (elliVM) [reply src/main/java/com/teragrep/pth10/steps/teragrep/TeragrepBloomStep.java]: Moved FilterTypes and LazyConnection to class variables, number of class variables are high for now but refactoring is easier with this change. 
Refactoring will also lower the number of variables of this class.

**Review by 51-code — APPROVED** [review]
LGTM now with the fixes.

**Review by eemhu — CHANGES_REQUESTED** [review]
rebase needed

**Comment by elliVM** [comment]
Will do a new PR as this one has diverged too much
